src/distributions/implicit.jl

```diff
@@ -0,0 +1,43 @@
+"""
+    ImplicitDistribution(sample_function, args...)
+
+Define a distribution that can only be sampled from using `rand`, but has no explicit `pdf`.
+
+Each time `rand(rng, d::ImplicitDistribution)` is called,
+```julia
+sample_function(args..., rng)
+```
+will be called to generate a new sample.
+
+`ImplicitDistribution` is designed to be used with anonymous functions or the `do` syntax as follows:
+
+# Examples
+
+```julia
+ImplicitDistribution(rng->rand(rng)^2)
+```
+
+```julia
+struct MyMDP <: MDP{Float64, Int} end
+
+function POMDPs.transition(m::MyMDP, s, a)
+    ImplicitDistribution(s, a) do s, a, rng
+        return s + a + 0.001*randn(rng)
+    end
+end
+
+td = transition(MyMDP(), 1.0, 1)
+rand(td) # will return a number near 2
+```
+"""
+struct ImplicitDistribution{F<:Function, A}
```

Comment:
You can make callable objects Functions e.g. `struct MyType <: Function`. Probably if someone passes an object here, it will be custom made for this purpose, or it is not too difficult to make a small closure. I think that the documentation/clarity bonus for including the constraint outweighs the downside. Am I forgetting anything?

Reply:
I guess your are right. I forgot that you could just do sub-type `Function`.

Reply:
Actually, I am going to remove the `Function` constraint. It is an antipattern https://www.juliabloggers.com/julialang-antipatterns/